wallet/wallet-controller/src/sync/mod.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::time::Duration;
+
+use chainstate::ChainInfo;
+use common::{
+    chain::{Block, ChainConfig, GenBlock},
+    primitives::{BlockHeight, Id},
+};
+use node_comm::node_traits::NodeInterface;
+use tokio::sync::mpsc;
+
+struct NextBlockInfo {
+    pub common_block_id: Id<GenBlock>,
+    pub block_id: Id<Block>,
+    pub block_height: BlockHeight,
+}
+
+pub struct FetchedBlock {
+    pub block: Block,
+    pub block_height: BlockHeight,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum FetchBlockError<T: NodeInterface> {
+    #[error("Unexpected RPC error: {0}")]
+    UnexpectedRpcError(T::Error),
+    #[error("Unexpected genesis block received at height {0}")]
+    UnexpectedGenesisBlock(BlockHeight),
+    #[error("There is no block at height {0}")]
+    NoBlockAtHeight(BlockHeight),
+    #[error("Block with id {0} not found")]
+    BlockNotFound(Id<Block>),
+    #[error("Invalid prev block id: {0}, expected: {1}")]
+    InvalidPrevBlockId(Id<GenBlock>, Id<GenBlock>),
+}
+
+pub type BlockFetchResult<T> = Result<FetchedBlock, FetchBlockError<T>>;
+
+pub async fn run_state_sync<T: NodeInterface>(state_tx: mpsc::Sender<ChainInfo>, rpc_client: T) {
+    let mut last_state = None;
+    while !state_tx.is_closed() {
+        let state_res = rpc_client.chainstate().await;
+        match state_res {
+            Ok(state) => {
+                if last_state.as_ref() != Some(&state) {
```

Comment:
It's better to just compare the tip hash. It's wasteful to have to compare everything.

Reply:
This check is negligible compared to the cost of the RPC call itself.

Reply:
Chainstate info can, in the future, contain things like relative time or whatever. It's not really safe to do it this way.

Reply:
Okay, updated, thought I still don't see what could possibly go wrong if we have up-to-date `ChainInfo`.